File: server/src/models/Conversation.ts
import mongoose, { Document, Schema, Types } from 'mongoose';

export type ConversationType = 'private' | 'group';
export type ConversationSource = 'internal' | 'whatsapp_business';

export interface IGroupMetadata {
  name: string;
  description?: string;
  image?: string;
  admins: Types.ObjectId[];
  createdBy: Types.ObjectId;
  inviteToken?: string;
}

export interface IWhatsAppMetadata {
  customerPhoneNumber: string;
  customerName?: string;
  businessPhoneNumberId: string;
  waConversationId?: string;
}

export interface IConversation extends Document {
  _id: Types.ObjectId;
  type: ConversationType;
  source: ConversationSource;
  participants: Types.ObjectId[];
  group?: IGroupMetadata;
  whatsappMetadata?: IWhatsAppMetadata;
  lastMessage?: Types.ObjectId;
  lastMessageAt?: Date;
  unreadCounts: Map<string, number>;
  pinnedBy: Types.ObjectId[];
  archivedBy: Types.ObjectId[];
  mutedBy: Types.ObjectId[];
  disappearingTimer: number;
  createdAt: Date;
  updatedAt: Date;
}

const GroupMetadataSchema = new Schema<IGroupMetadata>({
  name: { type: String, required: true, trim: true },
  description: { type: String, default: '' },
  image: { type: String, default: '' },
  admins: [{ type: Schema.Types.ObjectId, ref: 'User' }],
  createdBy: { type: Schema.Types.ObjectId, ref: 'User', required: true },
  inviteToken: { type: String },
}, { _id: false });

const WhatsAppMetadataSchema = new Schema<IWhatsAppMetadata>({
  customerPhoneNumber: { type: String, required: true },
  customerName: { type: String, default: '' },
  businessPhoneNumberId: { type: String, required: true },
  waConversationId: { type: String },
}, { _id: false });

const ConversationSchema = new Schema<IConversation>({
  type: { type: String, enum: ['private', 'group'], required: true, default: 'private' },
  source: { type: String, enum: ['internal', 'whatsapp_business'], required: true, default: 'internal' },
  participants: [{ type: Schema.Types.ObjectId, ref: 'User' }],
  group: { type: GroupMetadataSchema },
  whatsappMetadata: { type: WhatsAppMetadataSchema },
  lastMessage: { type: Schema.Types.ObjectId, ref: 'Message' },
  lastMessageAt: { type: Date, default: Date.now },
  unreadCounts: { type: Map, of: Number, default: () => new Map() },
  pinnedBy: [{ type: Schema.Types.ObjectId, ref: 'User' }],
  archivedBy: [{ type: Schema.Types.ObjectId, ref: 'User' }],
  mutedBy: [{ type: Schema.Types.ObjectId, ref: 'User' }],
  disappearingTimer: { type: Number, default: 0 },
}, {
  timestamps: true,
  toJSON: {
    transform(_doc, ret) {
      delete (ret as any).__v;
      return ret;
    }
  }
});

// Indexes for high performance querying
ConversationSchema.index({ participants: 1, updatedAt: -1 });
ConversationSchema.index({ 'whatsappMetadata.customerPhoneNumber': 1, 'whatsappMetadata.businessPhoneNumberId': 1 });
ConversationSchema.index({ 'group.inviteToken': 1 });

export const Conversation = mongoose.model<IConversation>('Conversation', ConversationSchema);

File: server/src/main/java/com/mchat/model/Conversation.java
package com.mchat.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.Instant;
import java.util.*;

@Entity
@Table(name = "conversations")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Conversation {

    @Id
    @Column(name = "id", length = 36, nullable = false, updatable = false)
    @JsonProperty("_id")
    private String id;

    @Column(nullable = false, length = 20)
    @Builder.Default
    private String type = "private"; // private, group

    @Column(nullable = false, length = 30)
    @Builder.Default
    private String source = "internal"; // internal, whatsapp_business

    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(
            name = "conversation_participants",
            joinColumns = @JoinColumn(name = "conversation_id"),
            inverseJoinColumns = @JoinColumn(name = "user_id")
    )
    @Builder.Default
    private List<User> participants = new ArrayList<>();

    // Group metadata
    private String groupName;

    @Column(length = 1000)
    private String groupDescription;

    @Column(length = 1000)
    private String groupImage;

    private String groupCreatedBy;

    private String groupInviteToken;

    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "conversation_admins", joinColumns = @JoinColumn(name = "conversation_id"))
    @Column(name = "admin_user_id")
    @Builder.Default
    private Set<String> groupAdminIds = new HashSet<>();

    // WhatsApp metadata
    private String waCustomerPhone;
    private String waCustomerName;
    private String waPhoneNumberId;
    private String waConversationId;

    // Last message info
    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "last_message_id")
    private Message lastMessage;

    private Instant lastMessageAt;

    @Builder.Default
    private Integer disappearingTimer = 0; // seconds

    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "conversation_pinned", joinColumns = @JoinColumn(name = "conversation_id"))
    @Column(name = "user_id")
    @Builder.Default
    private Set<String> pinnedBy = new HashSet<>();

    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "conversation_archived", joinColumns = @JoinColumn(name = "conversation_id"))
    @Column(name = "user_id")
    @Builder.Default
    private Set<String> archivedBy = new HashSet<>();

    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "conversation_muted", joinColumns = @JoinColumn(name = "conversation_id"))
    @Column(name = "user_id")
    @Builder.Default
    private Set<String> mutedBy = new HashSet<>();

    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "conversation_unread", joinColumns = @JoinColumn(name = "conversation_id"))
    @MapKeyColumn(name = "user_id")
    @Column(name = "unread_count")
    @Builder.Default
    private Map<String, Integer> unreadCounts = new HashMap<>();

    @CreationTimestamp
    private Instant createdAt;

    @UpdateTimestamp
    private Instant updatedAt;

    @PrePersist
    public void prePersist() {
        if (this.id == null || this.id.isBlank()) {
            this.id = UUID.randomUUID().toString();
        }
        if (this.lastMessageAt == null) {
            this.lastMessageAt = Instant.now();
        }
    }

    @JsonProperty("group")
    public Map<String, Object> getGroup() {
        if (!"group".equalsIgnoreCase(type) && groupName == null) {
            return null;
        }
        Map<String, Object> g = new HashMap<>();
        g.put("name", groupName != null ? groupName : "Group");
        g.put("description", groupDescription);
        g.put("image", groupImage);
        g.put("admins", groupAdminIds);
        g.put("createdBy", groupCreatedBy);
        g.put("inviteToken", groupInviteToken);
        return g;
    }

    @JsonProperty("whatsappMetadata")
    public Map<String, Object> getWhatsappMetadata() {
        if (!"whatsapp_business".equalsIgnoreCase(source) && waCustomerPhone == null) {
            return null;
        }
        Map<String, Object> wa = new HashMap<>();
        wa.put("customerPhoneNumber", waCustomerPhone);
        wa.put("customerName", waCustomerName);
        wa.put("businessPhoneNumberId", waPhoneNumberId);
        wa.put("waConversationId", waConversationId);
        return wa;
    }
}
